### PR TITLE
add ability to add viz to dashboard from query edit page

### LINKED ISSUE #955: Add a button / drop-down from the query window to quickly add it to a dashboard.

### PR DESCRIPTION
**Pull request by alison985**
Fixes #955. Ports this feature from Mozilla's fork.

Previous PR on Mozilla: https://github.com/mozilla/redash/pull/184
Previous code review and discussion: https://github.com/getredash/redash/pull/2656 (Closed because I keep messing up merge/rebase from upstream master and while I did I found an existing dashboard search implementation in getredash/redash so I modified this feature to use that instead.)

Add to Dashboard modal:
add-to-dashboard-modal

iPhone6s screen size:
iphone6s

MacBook Air 13: full window:
macbookair 13 inch full

### CODE REVIEW

File: client/app/pages/queries/view.js [hunk 1]
```diff
@@ -440,6 +440,19 @@ function QueryViewCtrl(
...
 
+  $scope.openAddToDashboardForm = (visId) => {
+    const visualization = getVisualization(visId);
+    $uibModal.open({
+      component: 'addToDashboardDialog',
+      size: 'sm',
+      resolve: {
+        query: $scope.query,
+        vis: visualization,
+        saveAddToDashboard: () => $scope.saveAddToDashboard,
```
Comment (arikfr) [inline client/app/pages/queries/view.js]: I couldn't find where this is used to where `$scope.saveAddToDashboard` is defined. Maybe a leftover?

File: client/app/pages/queries/add-to-dashboard.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      this.selected_query = this.resolve.query.id;
+      const widget = new Widget({
+        visualization_id: this.vis && this.vis.id,
+        dashboard_id: dash.id,
+        options: {},
+        width: 1,
+        type: 'visualization',
+      });
+      widget.save().then(() => {
+        this.selectedDashboard = Dashboard.get({ slug: dash.slug }, () => {});
```
Comment (arikfr) [inline client/app/pages/queries/add-to-dashboard.js]: Why is this needed?

File: client/app/pages/queries/add-to-dashboard.html [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<div class="modal-header">
+    <button type="button" class="close" aria-label="Close" ng-click="$ctrl.close()"><span aria-hidden="true">&times;</span></button>
+    <h4 class="modal-title">Add to Dashboard</h4>
+</div>
+<div class="modal-body">
+    <form name="alertForm" class="form">
+      <div class="form-group">
+        <!-- <label>Limit Dashboard search to dashboards I've created? </label>
+        <input type="checkbox" ng-model="$ctrl.limitToUsersDashboards" on-change="$ctrl.dashboardList == $select.search"   ng-disabled="$ctrl.saveInProgress" /> -->
```
Comment (arikfr) [inline client/app/pages/queries/add-to-dashboard.html]: Remove?

**Comment by arikfr** [comment]
@kravets-levko FYI: we will need to support the new flow from #2641 here too. The easiest way to do this is to be able to use the component/modal you implement for #2641 in this modal as well. Only change being is that the first step will be to choose a dashboard instead of choosing query/visualization.

---

The bottom section becomes crowded:

> iPhone6s screen size:
iphone6s

> MacBook Air 13: full window:
macbookair 13 inch full

My suggestion:

* Replace {Download, Add to Dashboard, Embed} buttons with a single [Kebab menu](https://thenounproject.com/term/kebab-menu/198754/) that will have these options.
* Hide Add to Dashboard and Edit Visualization on small screen sizes.

@kocsmy @alison985 WDYT?

@alison985 you don't need to implement these changes as part of this pull request (unless you want to), but without them this pull request will need to wait for V5 to finalize before merging into master.

**Comment by kocsmy** [comment]
Hiding these buttons behind a Kebab menu would make the interface lighter for sure. I'm wondering if that's fine and we're not hiding important features?

Also the download which has 2 option now should just open a modal for selecting the type, instead of further complicating the Kebab menu with submenus.

**Comment by kravets-levko** [comment]
@kocsmy 

> Also the download which has 2 option now should just open a modal for selecting the type, instead of further complicating the Kebab menu with submenus.

Not necessarily: it could be two "Download ..." options in plain menu, like this:
```
Item 1
Item 2
Item 3
--------------------
Download as CSV file
Download as Excel file
--------------------
One more menu item
```

**Comment by kocsmy** [comment]
Thanks, @kravets-levko that's also a good idea!

File: client/app/pages/queries/add-to-dashboard.html (hunk 3 above)
Comment (alison985) [reply client/app/pages/queries/add-to-dashboard.html]: Yes, all 3 of these sections are removed in the next commit.

**Comment by alison985** [comment]
I've implemented the kebab menu. On an iPhone 6s it looks like this:
screenshot 2018-09-09 12 31 07

The problem is that there isn't enough vertical height for the full menu now. The Add to Dashboard option gets cut off.
screenshot 2018-09-09 12 32 38

I think the solution is to reduce the amount of screen space the query textarea takes up. I'm just not sure the best way to implement that.

Also, I'm not sure what you mean about the #2641 flow integration so I'm going to leave that part alone.

**Comment by jezdez** [comment]
@alison985 Still needs a rebase. 
o5x49tdgjzo1ru8q831vuevdznmwnr3p-medium
